packages/firestore/test/unit/specs/bundle_spec.test.ts

```diff
@@ -0,0 +1,357 @@
+/**
+ * @license
+ * Copyright 2017 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Query } from '../../../src/core/query';
+import { doc, path, TestSnapshotVersion, version } from '../../util/helpers';
+
+import { describeSpec, specTest } from './describe_spec';
+import { client, spec } from './spec_builder';
+import { TestBundleBuilder } from '../../util/bundle_data';
+import {
+  JSON_SERIALIZER,
+  TEST_DATABASE_ID
+} from '../local/persistence_test_helpers';
+import { DocumentKey } from '../../../src/model/document_key';
+import * as api from '../../../src/protos/firestore_proto_api';
+import { Value } from '../../../src/protos/firestore_proto_api';
+import { toVersion } from '../../../src/remote/serializer';
+
+interface TestBundleDocument {
+  key: DocumentKey;
+  readTime: TestSnapshotVersion;
+  createTime?: TestSnapshotVersion;
+  updateTime?: TestSnapshotVersion;
+  content?: api.ApiClientObjectMap<Value>;
+}
+function bundleWithDocument(testDoc: TestBundleDocument): string {
+  const builder = new TestBundleBuilder(TEST_DATABASE_ID);
+  builder.addDocumentMetadata(
+    testDoc.key,
+    toVersion(JSON_SERIALIZER, version(testDoc.readTime)),
+    !!testDoc.createTime
+  );
+  if (testDoc.createTime) {
+    builder.addDocument(
+      testDoc.key,
+      toVersion(JSON_SERIALIZER, version(testDoc.createTime)),
+      toVersion(JSON_SERIALIZER, version(testDoc.updateTime!)),
+      testDoc.content!
+    );
+  }
+  return builder.build(
+    'test-bundle',
+    toVersion(JSON_SERIALIZER, version(testDoc.readTime))
+  );
+}
+
+describeSpec('Bundles:', [], () => {
+  specTest('Newer docs from bundles should overwrite cache.', [], () => {
+    const query1 = Query.atPath(path('collection'));
+    const docA = doc('collection/a', 1000, { key: 'a' });
+    const docAChanged = doc('collection/a', 2999, { key: 'b' });
+
+    const bundleString = bundleWithDocument({
+      key: docA.key,
+      readTime: 3000,
+      createTime: 1999,
+      updateTime: 2999,
+      content: { key: { stringValue: 'b' } }
+    });
+
+    return spec()
+      .userListens(query1)
+      .watchAcksFull(query1, 1000, docA)
+      .expectEvents(query1, { added: [docA] })
+      .loadBundle(bundleString)
+      .expectEvents(query1, { modified: [docAChanged] });
```

Comment:
Should this be raised with `fromCache: true`?

Reply:
Do you want to add a comment here that links to the bug you filed? Makes it easier to parse this test.